### DISCUSSION TITLE
React useSelector vs useMemo and what are the case you should use useSelector ?

### DISCUSSION THREAD

**Discussion by maxime4000** [statelyai/xstate#2119] [Q&A] [answered]
What is the difference between useSelector and useMemo ? 
```.jsx
const [state, send, service] = useMachine(fetchMachine);
// useSelector
const datas1 = useSelector(service, ((state) => state.context.datas.filter(someFilter)))

// useMemo
const datas2 = useMemo(() => {
  return state.context.datas.filter(someFilter);
}, [state.context.datas, someFilter]);
```
So yeah, it's not the best example I could gave, but I don't understand what useSelector bring to the table. Does it has better performance for comparing previous/next value, or is it when you only have the service ? Or maybe it's best when you use other value than state.context ? Or maybe I'm not using xstate right when doing useMemo? 

Maybe I'm not understanding selector, because I'm not understanding Service and Actor ?
My main machine is inside of a React.Context and the hook return an object that hold everything I could need and I'm destructuring what I need in my component.  
```.tsx
{
  state, 
  send, 
  service, 
  context: state.context, 
  // init => assign({fetchXMachine: (ctx,event) => spawn(fetchXMachine),})
  fetchXContext: state.context.fetchXMachine?.state.context ?? ({} as FetchXMachineContext), 
  ...
}
```
Probably a bad design, but it work.
Now the real question: Would that be a good case to remove fetchXContext in favor of useService and/or useSelector for component who need some value in those children context ?
```.jsx
const { service } = useMachineInReactContext();
const [state, send] = useService(useSelector(service, (state) => state.children.fetchXMachine));
const fetchXContext = state.context; 
```
Using xstate in react is like never knowing if the design is good or not, but at least it always work.

**Comment by davidkpiano** [comment, marked as answer]
In this example:

```js
const [state, send, service] = useMachine(fetchMachine);
// useSelector
const datas1 = useSelector(service, ((state) => state.context.datas.filter(someFilter)))

// useMemo
const datas2 = useMemo(() => {
  return state.context.datas.filter(someFilter);
}, [state.context.datas, someFilter]);
```

You're not going to get any performance benefits.

_However_, if you only have the `service`, then this:

```js
const Thing = ({ service }) => {
  const [state, send] = useService(service);
  const count = state.context.count; // doesn't matter if you useMemo() here or not
});
```

is going to rerender **every time the `state` changes**, whereas this:

```js
const Thing = ({ service }) => {
  const count = useSelector(service, state => state.context.count);
});
```

is going to rerender **only when `count` changes**.

Hopefully that helps!
